R/pareto_smooth.R

```diff
@@ -0,0 +1,326 @@
+
+  convergence_rate <- ps_convergence_rate(k, S)
+
+  other_diags <- list(
+    "min_ss" = min_ss,
+    "khat_threshold" = khat_threshold,
+    "convergence_rate" = convergence_rate
+  )
+}
+
+#' Pareto-smoothing minimum sample-size
+#'
+#' Given Pareto-k computes the minimum sample size for reliable Pareto
+#' smoothed estimate (to have small probability of large error)
+#' @param k pareto k value
+#' @param ... unused
+#' @return minimum sample size
+ps_min_ss <- function(k, ...) {
+  # Eq (11) in PSIS paper
+  10^(1 / (1 - max(0, k)))
+}
+
+#' Pareto-smoothing k-hat threshold
+#'
+#' Given sample size S computes khat threshold for reliable Pareto
+#' smoothed estimate (to have small probability of large error). See
+#' section 3.2.4, equation (13).
+#' @param S sample size
+#' @param ... unused
+#' @return threshold
+ps_khat_threshold <- function(S, ...) {
+  1 - 1 / log10(S)
+}
+
+#' Pareto-smoothing convergence rate
+#'
+#' Given S and scalar or array of k's, compute the relative
+#' convergence rate of PSIS estimate RMSE
+#' @param k pareto-k values
+#' @param S sample size
+#' @param ... unused
+#' @return convergence rate
+ps_convergence_rate <- function(k, S, ...) {
+  # allow array of k's
+  rate <- numeric(length(k))
+  # k<0 bounded distribution
+  rate[k < 0] <- 1
+  # k>0 non-finite mean
+  rate[k > 1] <- 0
+  # limit value at k=1/2
+  rate[k == 0.5] <- 1 - 1 / log(S)
+  # smooth approximation for the rest (see Appendix of PSIS paper)
+  ki <- (k > 0 & k < 1 & k != 0.5)
+  kk <- k[ki]
+  rate[ki] <- pmax(
+    0,
+    (2 * (kk - 1) * S^(2 * kk + 1) + (1 - 2 * kk) * S^(2 * kk) + S^2) /
+      ((S - 1) * (S - S^(2 * kk)))
+  )
+  rate
+}
+
+ps_tail_length <- function(S, r_eff, ...) {
+  # Appendix H in PSIS paper
+  ifelse(S > 225, ceiling(3 * sqrt(S / r_eff)), S / 5)
+}
+
+#' Pareto-k diagnostic message
+#'
+#' Given S and scalar and k, form a diagnostic message string
+#' @param diags (numeric) named vector of diagnostic values
+#' @param ... unused
+#' @return diagnostic message
+pareto_k_diagmsg <- function(diags, ...) {
+  khat <- diags$khat
+  min_ss <- diags$min_ss
+  khat_threshold <- diags$khat_threshold
+  convergence_rate <- diags$convergence_rate
+  msg <- NULL
+  if (khat > 1) {
+    msg <- paste0(msg,'All estimates are unreliable. If the distribution of ratios is bounded,\n',
+                  'further draws may improve the estimates, but it is not possible to predict\n',
+                  'whether any feasible sample size is sufficient.')
+  } else {
+    if (khat > khat_threshold) {
+      msg <- paste0(msg, 'S is too small, and sample size larger than ', round(min_ss, 0), ' is neeeded for reliable results.\n')
+    } else {
+      msg <- paste0(msg, 'To halve the RMSE, approximately ', round(2^(2/convergence_rate),1), ' times bigger S is needed.\n')
+    }
+    if (khat > 0.7) {
+      msg <- paste0(msg, 'Bias dominates RMSE, and the variance based MCSE is underestimated.\n')
+    }
+  }
+  message(msg)
```

Comment:
this function should return the main argument invisibly, invisible(diags), or invisible(NULL). The message is printed to terminal anyway. Or do you really want to return the message itself?

Reply:
I added a return of `invisible(diags)` to this.